lib/routes.js

```diff
@@ -145,29 +148,85 @@ module.exports = async function (app, opts) {
     verifyClient,
     lruGatewayResolvers,
     entityResolversFactory,
-    onlyPersisted,
-    persistedQueries = {},
+    persistedQuerySettings,
     allowBatchedQueries
   } = opts
 
+  // Load the persisted query settings
+  const {
+    isPersistedQuery,
+    getHash,
+    getQueryFromHash,
+    getHashForQuery,
+    saveQuery,
+    notFoundError,
+    notSupportedError
+  } = persistedQuerySettings || {}
+
+  const enablePersistedQueries = !!persistedQuerySettings
+
   async function execute (body, request, reply) {
     let { query } = body
-    const { persisted, operationName, variables } = body
+    const { operationName, variables } = body
+
+    // Generate the context for this request
+    let context = {}
+    if (contextFn) {
+      context = await contextFn(request, reply)
+    }
+
+    // Verify if a query matches the persisted format, if required
+    let persisted
+    if (enablePersistedQueries) {
+      persisted = isPersistedQuery(body)
+      if (persisted) {
+        // This is a peristed query, so we use the hash in the request
+        // to load the full query string.
+
+        // Extract the hash from the request
+        const hash = getHash && getHash(body)
+        if (hash) {
+          // Load the query for the provided hash
+          query = getQueryFromHash && await getQueryFromHash(hash)
```

Comment:
I think we should validate during plugin setup that if somebody is enabling persisted queries they should pass in a `getQueryFromHash` function.